packages/remix-react/routes.tsx

```diff
@@ -183,82 +125,56 @@ async function loadRouteModuleWithBlockingLinks(
 }
 
-function createLoader(route: EntryRoute, routeModules: RouteModules) {
-  let loader: ClientRoute["loader"] = async ({ url, signal, submission }) => {
-    if (route.hasLoader) {
-      let routeModulePromise = loadRouteModuleWithBlockingLinks(
-        route,
-        routeModules
-      );
-      try {
-        let result = await fetchData(url, route.id, signal, submission);
-
-        if (result instanceof Error) throw result;
-
-        /* TODO: Handle client side redirects
-        let redirect = await checkRedirect(result);
-        if (redirect) return redirect;
-        */
-
-        if (isCatchResponse(result)) {
-          throw new CatchValue(
-            result.status,
-            result.statusText,
-            await extractData(result)
-          );
-        }
-
-        return extractData(result);
-      } finally {
-        await routeModulePromise;
-      }
-    } else {
-      await loadRouteModuleWithBlockingLinks(route, routeModules);
-    }
-  };
-
-  return loader;
-}
-
-function createAction(route: EntryRoute, routeModules: RouteModules) {
-  let action: ClientRoute["action"] = async ({ url, signal, submission }) => {
-    let routeModulePromise = await loadRouteModuleWithBlockingLinks(
+function createDataFunction(
+  route: EntryRoute,
+  routeModules: RouteModules,
+  isAction: boolean
+): LoaderFunction | ActionFunction {
+  return async ({ request }) => {
+    let routeModulePromise = loadRouteModuleWithBlockingLinks(
       route,
       routeModules
     );
-
     try {
-      if (!route.hasAction) {
+      if (isAction && !route.hasAction) {
         console.error(
           `Route "${route.id}" does not have an action, but you are trying ` +
             `to submit to it. To fix this, please add an \`action\` function to the route`
         );
+        // TODO: Should we `return null` here like we do for loaders?  Is there
+        // a benefit to triggering a network request that will error/404?
+      } else if (!isAction && !route.hasLoader) {
+        return null;
       }
 
-      let result = await fetchData(url, route.id, signal, submission);
+      let result = await fetchData(request, route.id, isAction);
 
       if (result instanceof Error) {
         throw result;
       }
 
-      /* TODO: Handle client side redirects
-      let redirect = await checkRedirect(result);
-      if (redirect) return redirect;
-      */
+      if (isRedirectResponse(result)) {
+        throw getRedirect(result);
+      }
 
       if (isCatchResponse(result)) {
-        throw new CatchValue(
-          result.status,
-          result.statusText,
-          await extractData(result)
-        );
+        throw result;
       }
 
-      return extractData(result);
+      return result;
     } finally {
       await routeModulePromise;
     }
   };
+}
 
-  return action;
+function getRedirect(response: Response): Response {
+  let status = parseInt(response.headers.get("X-Remix-Status")!, 10) || 302;
```

Comment:
This is new behavior to support the bug fix in the router that respects 307/308s and preserves the method/body on the redirect: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/307

Remix was previously not sending the status code along at all because it became a 204